### PR TITLE
feat: parse some fields from SGX extension

### PR DESCRIPTION
**Pull request by rjzak**
Closes: #113

Cleans up some SGX extension parsing in:
* https://github.com/profianinc/steward/pull/173
* https://github.com/enarx/enarx/pull/2434

Limitations:
* The SGX extensions have a lot of OID: Value pairs and I'm not sure this is the best way to handle them, since this isn't checking the OIDs. @npmccallum thoughts?
* [X] ~~Haven't yet tested on PCK from a multi-CPU system, which will have five additional fields.~~

Example:
* Single-CPU: https://lapo.it/asn1js/#MIIBwDAeBgoqhkiG-E0BDQEBBBCUXkyhQ6F4Vd-P0qfZEkzyMIIBYwYKKoZIhvhNAQ0BAjCCAVMwEAYLKoZIhvhNAQ0BAgECAQYwEAYLKoZIhvhNAQ0BAgICAQYwEAYLKoZIhvhNAQ0BAgMCAQIwEAYLKoZIhvhNAQ0BAgQCAQIwEAYLKoZIhvhNAQ0BAgUCAQIwEAYLKoZIhvhNAQ0BAgYCAQEwEAYLKoZIhvhNAQ0BAgcCAQQwEAYLKoZIhvhNAQ0BAggCAQAwEAYLKoZIhvhNAQ0BAgkCAQAwEAYLKoZIhvhNAQ0BAgoCAQAwEAYLKoZIhvhNAQ0BAgsCAQAwEAYLKoZIhvhNAQ0BAgwCAQAwEAYLKoZIhvhNAQ0BAg0CAQAwEAYLKoZIhvhNAQ0BAg4CAQAwEAYLKoZIhvhNAQ0BAg8CAQAwEAYLKoZIhvhNAQ0BAhACAQAwEAYLKoZIhvhNAQ0BAhECAQowHwYLKoZIhvhNAQ0BAhIEEAYGAgICAQQAAAAAAAAAAAAwEAYKKoZIhvhNAQ0BAwQCAAAwFAYKKoZIhvhNAQ0BBAQGAHBuRwAAMA8GCiqGSIb4TQENAQUKAQA
* Multi-CPU ("Platform"): https://lapo.it/asn1js/#MIICKDAeBgoqhkiG-E0BDQEBBBBRg_x7iwvN-wThN1uhLINFMIIBZQYKKoZIhvhNAQ0BAjCCAVUwEAYLKoZIhvhNAQ0BAgECAQQwEAYLKoZIhvhNAQ0BAgICAQQwEAYLKoZIhvhNAQ0BAgMCAQMwEAYLKoZIhvhNAQ0BAgQCAQMwEQYLKoZIhvhNAQ0BAgUCAgD_MBEGCyqGSIb4TQENAQIGAgIA_zAQBgsqhkiG-E0BDQECBwIBADAQBgsqhkiG-E0BDQECCAIBADAQBgsqhkiG-E0BDQECCQIBADAQBgsqhkiG-E0BDQECCgIBADAQBgsqhkiG-E0BDQECCwIBADAQBgsqhkiG-E0BDQECDAIBADAQBgsqhkiG-E0BDQECDQIBADAQBgsqhkiG-E0BDQECDgIBADAQBgsqhkiG-E0BDQECDwIBADAQBgsqhkiG-E0BDQECEAIBADAQBgsqhkiG-E0BDQECEQIBCzAfBgsqhkiG-E0BDQECEgQQBAQDA___AAAAAAAAAAAAADAQBgoqhkiG-E0BDQEDBAIAADAUBgoqhkiG-E0BDQEEBAYAYGoAAAAwDwYKKoZIhvhNAQ0BBQoBATAeBgoqhkiG-E0BDQEGBBBQc-trHNpacEyKD8Ei4N-6MEQGCiqGSIb4TQENAQcwNjAQBgsqhkiG-E0BDQEHAQEB_zAQBgsqhkiG-E0BDQEHAgEB_zAQBgsqhkiG-E0BDQEHAwEB_w

Documentation: https://download.01.org/intel-sgx/sgx-dcap/1.10.3/linux/docs/SGX_PCK_Certificate_CRL_Spec-1.4.pdf

### CODE REVIEW

File: src/pck/mod.rs [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        }
+
+        for extension in extensions.iter() {
+            if extension.extn_id == SGX_EXTENSION {
+                let sgx_extension: SgxExtensionRaw = match Decode::from_der(extension.extn_value) {
+                    Ok(s) => s,
+                    Err(e) => {
+                        return Err(SgxExtensionError::DerDecodingError(e));
+                    }
+                };
```
Comment (haraldh) [inline src/pck/mod.rs:52]: use `map_err(...)?`

File: src/pck/mod.rs [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        match self {
+            SgxExtensionError::NoExtensionsFound => write!(f, "SGX: No extensions found"),
+            SgxExtensionError::MissingSgxExtension => write!(f, "SGX: Missing extension"),
+            SgxExtensionError::DerDecodingError(e) => write!(f, "SGX: Der decoding error: {e}"),
+        }
+    }
+}
+
+impl<'a> SgxExtension<'a> {
+    pub fn from_x509_extension(extensions: &'a Extensions) -> Result<Self, SgxExtensionError> {
```
Comment (haraldh) [inline src/pck/mod.rs:40]: from_x509_extensions? plural?

Comment (rjzak) [reply src/pck/mod.rs:40]: Yes, because the Certificate has multiple extensions, so we receive all of them, and find the one which is the SGX extension.

Comment (haraldh) [reply src/pck/mod.rs:40]: what I meant is: please change the function name to have extensions in plural

Comment (rjzak) [reply src/pck/mod.rs]: Fixed!

File: src/pck/mod.rs (hunk 1 above)
Comment (rjzak) [reply src/pck/mod.rs]: Done

File: src/pck/mod.rs [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+
+mod raw;
+
+use raw::SgxExtensionRaw;
+
+use const_oid::ObjectIdentifier;
+use der::Decode;
+use x509::ext::Extensions;
+
+const SGX_EXTENSION: ObjectIdentifier = ObjectIdentifier::new_unwrap("1.2.840.113741.1.13.1");
```
Comment (haraldh) [inline src/pck/mod.rs:11]: use `SgxExtensionRaw::OID` instead of this?

File: src/pck/mod.rs [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+            SgxExtensionError::MissingSgxExtension => write!(f, "SGX: Missing extension"),
+            SgxExtensionError::DerDecodingError(e) => write!(f, "SGX: Der decoding error: {e}"),
+        }
+    }
+}
+
+impl<'a> SgxExtension<'a> {
+    pub fn from_x509_extensions(extensions: &'a Extensions) -> Result<Self, SgxExtensionError> {
+        if extensions.is_empty() {
+            return Err(SgxExtensionError::NoExtensionsFound);
```
Comment (haraldh) [inline src/pck/mod.rs:42]: could also be the  Missing... or vice versa... why not combine both?

File: src/pck/mod.rs [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        }
+
+        for extension in extensions.iter() {
```
Comment (haraldh) [inline src/pck/mod.rs:45]: is `extensions.iter().find(...)` better?

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/pck/mod.rs (hunk 4 above)
Comment (rjzak) [reply src/pck/mod.rs]: Done

File: src/pck/mod.rs (hunk 3 above)
Comment (rjzak) [reply src/pck/mod.rs]: Done

File: src/pck/mod.rs (hunk 5 above)
Comment (rjzak) [reply src/pck/mod.rs]: Wow, that's a decent amount of code reduction. I didn't know about `.iter().find()`

File: src/pck/mod.rs [hunk 6]
```diff
@@ -0,0 +1,114 @@
+// SPDX-License-Identifier: Apache-2.0
+
```
Comment (haraldh) [inline src/pck/mod.rs:2]:
```suggestion
///! SGX PCK Certificate parsing
///! 
///! see 1.3.5 Intel® SGX PCK Certificate of https://download.01.org/intel-sgx/sgx-dcap/1.10.3/linux/docs/SGX_PCK_Certificate_CRL_Spec-1.4.pdf
```

File: src/pck/raw.rs [hunk 7]
```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: Apache-2.0
+
```
Comment (haraldh) [inline src/pck/raw.rs:2]:
```suggestion
///! SGX PCK Certificate parsing
///! 
///! see 1.3.5 Intel® SGX PCK Certificate of https://download.01.org/intel-sgx/sgx-dcap/1.10.3/linux/docs/SGX_PCK_Certificate_CRL_Spec-1.4.pdf
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/pck/mod.rs (hunk 6 above)
Comment (rjzak) [reply src/pck/mod.rs]: Added

File: src/pck/raw.rs (hunk 7 above)
Comment (rjzak) [reply src/pck/raw.rs]: Added

File: src/pck/raw.rs [hunk 8]
```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: Apache-2.0
+///! SGX PCK Certificate parsing
+///!
+///! see 1.3.5 Intel® SGX PCK Certificate of https://download.01.org/intel-sgx/sgx-dcap/1.10.3/linux/docs/SGX_PCK_Certificate_CRL_Spec-1.4.pdf
```
Comment (haraldh) [inline src/pck/raw.rs:4]:
```suggestion

//! SGX PCK Certificate parsing
//!
//! see 1.3.5 Intel® SGX PCK Certificate of https://download.01.org/intel-sgx/sgx-dcap/1.10.3/linux/docs/SGX_PCK_Certificate_CRL_Spec-1.4.pdf

```

File: src/pck/mod.rs [hunk 9]
```diff
@@ -0,0 +1,116 @@
+// SPDX-License-Identifier: Apache-2.0
+///! SGX PCK Certificate parsing
+///!
+///! see 1.3.5 Intel® SGX PCK Certificate of https://download.01.org/intel-sgx/sgx-dcap/1.10.3/linux/docs/SGX_PCK_Certificate_CRL_Spec-1.4.pdf
```
Comment (haraldh) [inline src/pck/mod.rs:4]:
```suggestion

//! SGX PCK Certificate parsing
//!
//! see 1.3.5 Intel® SGX PCK Certificate of https://download.01.org/intel-sgx/sgx-dcap/1.10.3/linux/docs/SGX_PCK_Certificate_CRL_Spec-1.4.pdf

```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/pck/raw.rs (hunk 8 above)
Comment (rjzak) [reply src/pck/raw.rs]: Fixed

File: src/pck/mod.rs (hunk 9 above)
Comment (rjzak) [reply src/pck/mod.rs]: Fixed

**Review by haraldh — APPROVED** [review]

**Pull request merged**